packages/metro-config/src/ExpoMetroConfig.ts

```diff
@@ -0,0 +1,70 @@
+import path from 'path';
+
+import { getConfig, resolveModule } from '@expo/config';
+import { Reporter } from 'metro';
+import { ConfigT, InputConfigT, loadConfig } from 'metro-config';
+
+const INTERNAL_CALLSITES_REGEX = new RegExp(
+  [
+    '/Libraries/Renderer/implementations/.+\\.js$',
+    '/Libraries/BatchedBridge/MessageQueue\\.js$',
+    '/Libraries/YellowBox/.+\\.js$',
+    '/Libraries/LogBox/.+\\.js$',
+    '/Libraries/Core/Timers/.+\\.js$',
+    '/node_modules/react-devtools-core/.+\\.js$',
+    '/node_modules/react-refresh/.+\\.js$',
+    '/node_modules/scheduler/.+\\.js$',
+  ].join('|')
+);
+
+export function getDefaultConfig(projectRoot: string): InputConfigT {
+  const { exp } = getConfig(projectRoot, { skipSDKVersionRequirement: true });
+  const reactNativePath = resolveModule('react-native', projectRoot, exp);
+
+  return {
+    resolver: {
+      resolverMainFields: ['react-native', 'browser', 'main'],
+      platforms: ['ios', 'android', 'native'],
+    },
+    serializer: {
+      getModulesRunBeforeMainModule: () => [
+        require.resolve(path.join(reactNativePath, 'Libraries/Core/InitializeCore')),
+        // TODO: Bacon: load Expo side-effects
+      ],
+      getPolyfills: () => require(path.join(reactNativePath, 'rn-get-polyfills'))(),
+    },
+    symbolicator: {
+      customizeFrame: (frame: { file: string | null }) => {
+        const collapse = Boolean(frame.file && INTERNAL_CALLSITES_REGEX.test(frame.file));
+        return { collapse };
+      },
+    },
+    transformer: {
+      babelTransformerPath: require.resolve('../transformer'),
+      // TODO: Bacon: Add path for web platform
+      assetRegistryPath: path.join(reactNativePath, 'Libraries/Image/AssetRegistry'),
+      assetPlugins: [resolveModule('expo/tools/hashAssetFiles', projectRoot, exp)],
+    },
+  };
+}
+
+export interface LoadOptions {
+  config?: string;
+  maxWorkers?: number;
+  port?: number;
+  reporter?: Reporter;
+  resetCache?: boolean;
+  // TODO(ville): make this a part of the default config?
+  sourceExts?: string[];
```

Comment:
yeah we should do this. our current approach of passing in sourceExts in flag will clobber any value provided in metro.config.js, and people need to pass it in to packageOpts in app.json, which we should remove in favor of telling people to just use metro.config.js.

one tricky thing here is that the sourceExts need to be different depending on whether we are starting the packager for managed or bare workflow -- we don't use .expo.js in bare

Reply:
That is true. It needs to be possible to override it in `metro.config.js` and the interface for doing that could be something like this:
```js
// metro.config.js
const { getDefaultConfig } = require('@expo/metro-config');

const defaultConfig = getDefaultConfig(__dirname);

module.exports = {
  resolver: {
    sourceExts: [...defaultConfig.resolver.sourceExts, "svg"]
  }
};
```

This is indeed made a bit trickier by the fact that the default is different between bare and managed. We have access to the app configuration in `getDefaultConfig`, so we should be able to tell if it's a managed or bare app, *but* now it's also possible to override that setting for bare apps using the `--target` flag, at runtime: https://github.com/expo/expo-cli/pull/1734.

We might have to add an environment variable like `EXPO_WORKFLOW` or `EXPO_TARGET` to be able to access this information when `getDefaultConfig` is called from `metro.config.js`.

Or as an alternative, it kind of seems that ideally we wouldn't bake the `expo` "platform extension" into `sourceExts` at all (`expo.js`, `expo.jsx`, `expo.ts`, `expo.tsx`...), , but instead it would be separate, like the `resolver.platforms` config in Metro is separate from `sourceExts`. Then you would, regardless of workflow, simply have something like
```
sourceExts: ['ts', 'tsx', 'js', 'jsx', 'json', 'wasm']
```
and it would be possible to prefix each of them with `expo` for the managed target. To enable this, we might have to define a `CustomResolver` (the `resolve.resolveRequest` setting) or we could see if defining platforms like this would work:
```
platforms: ['expo.android', 'expo.ios', 'expo.web', 'android', 'ios', 'web']
```

Reply:
I'll open a new pull request for this change, once it's ready.